README.md
# cz-github-jira-conventional-footer

**cz-github-jira-conventional-footer** is a plugin for the [**commitizen tools**][commitizen], a toolset that helps you to create [**conventional commit messages**][conventional-commit]. Since the structure of conventional commits messages is standardized they are machine readable and allow commitizen to automaticially calculate and tag [**semantic version numbers**][semver] as well as create **CHANGELOG.md** files for your releases.

This plugin extends the commitizen tools by:
- **adding Jira issue identifiers** in a customisable footer of the commit message
- **creating links to GitHub** commits in the CHANGELOG.md
- **creating links to Jira** issues in the CHANGELOG.md

## Installation

Install with pip
```console
python -m pip install cz-github-jira-conventional-footer
```

## Configuration

The behavior of the plugin can be modified by adding the following values to your .yaml/.toml/.json config file.
Note that `github_repo` and `jira_base_url` are *required* while the others are _optional_.

| Parameter | Type | Required | Default | Description |
|---|---|---|---|---|
| `github_repo` | `str` | YES |  | Name of the repository on GitHub in format "`{username}`/`{repo name}`" |
| `jira_base_url` | `str` | YES |  | Base URL of the Jira application that tracks issues for this repository |
| `jira_prefix` | `str` | NO | None | Key for the Jira project that tracks issues for this repository. If issues for this repository are tracked on multiple Jira projects, this config value should be left blank which will prompt commitizen for full issue identifiers  |
| `jira_token` | `str` | NO | `Jira` | Token to use for the footer containing Jira issues. Cannot contain spaces or the phrase "Breaking Change". Can optionally specify the separator as either `:<space>`[^yaml-note] or `<space>#` by including it in the value. If the separator is not specified, a `:<space>` separator will automatically be added |
| `bump_pattern` | `str` | NO | [`commitizen.defaults.bump_pattern`][cz-defaults-bump-pattern] | Regex to extract information from commit (subject and body) |
| `bump_map` | `dict` | NO | [`commitizen.defaults.bump_map`][cz-defaults-bump-map] | Dictionary mapping the extracted information to a SemVer increment type (MAJOR, MINOR, PATCH) |
| `changelog_pattern` | `str` | NO | `bump_pattern` | Regex to validate the commits, this is useful to skip commits that don't meet your ruling standards like a Merge. Usually the same as bump_pattern |
| `change_type_map` | `dict` | NO | `{"feat":"Features", "fix":"Bug Fixes", "refactor":"Refactor", "perf":"Performance Improvements"}` | Convert the title of the change type that will appear in the changelog, if a value is not found, the original will be used |

[^yaml-note]: The colon (`:`) character has a special meaning in YAML syntax and must be escaped by enclosing the entire value in single quotes (ex: `jira_token: 'resolves-issue: '`)

### pre-commit

Add this plugin to the dependencies of your commit message linting with [pre-commit][pre-commit]. 

Example `.pre-commit-config.yaml` file:
```yaml
repos:
  - repo: https://github.com/commitizen-tools/commitizen
    rev: v2.17.13
    hooks:
      - id: commitizen
        additional_dependencies: [cz-github-jira-conventional-footer]
```
Install the hook with 
```bash
pre-commit install --hook-type commit-msg
```

## Example Usage

### Configuration

Example `.cz.yaml` file:
```yaml
commitizen:
  name: cz_github_jira_conventional_footer
  tag_format: v$version
  version: 0.3.1
  github_repo: brianburwell11/cz-github-jira-conventional-footer
  jira_base_url: https://myproject.atlassian.net
  jira_prefix: XX-
```

### Making a Commit

When you call commitizen `commit` you will be prompted to enter the scope of your commit as a Jira issue id (or multiple issue ids, prefixed or without prefix, see [config](#configuration) above).

```console
$ cz commit
? Select the type of change you are committing feat: A new feature. Correlates with MINOR in SemVer
? What is the scope of this change? (class or file name): (press [enter] to skip)
 changelog
? Write a short and imperative summary of the code changes: (lower case and no period)
 add GitHub and Jira badges to CHANGELOG
? Provide additional contextual information about the code changes: (press [enter] to skip)
 Modify `changelog_message_builder_hook()` to show the commit hash and Jira issues as badges in each entry of the changelog. This is enabled by named groups in the `commit_parser` regex pattern.
? Is this a BREAKING CHANGE? Correlates with MAJOR in SemVer No
? JIRA issue number (multiple "42, 123"). XX- 03, 04
? Add any additional footers. (press [enter] to skip)
 added-by: Brian Burwell <brianburwell11@gmail.com>
```

Creates the following commit message:
```
feat(changelog): add GitHub and Jira badges to CHANGELOG

Modify `changelog_message_builder_hook()` to show the commit hash and Jira issues as badges in each entry of the changelog. This is enabled by named groups in the `commit_parser` regex pattern.

Jira: XX-03,XX-04
added-by: Brian Burwell <brianburwell11@gmail.com>
```

### Reading the Changelog

The changelog created by commitizen (`cz changelog` or `cz bump --changelog`) contains links to the commits in Github and the Jira issues.
```markdown
### Features

- [![b56ea](https://img.shields.io/badge/-b56ea-%23121011.svg?style=flat-square&logo=github&logoColor=white)](https://github.com/brianburwell11/cz-github-jira-conventional-footer/commit/b56ea677af8a969614b159316d9d28a3267f7169)[![XX-03](https://img.shields.io/badge/-XX--03-dfe1e5.svg?style=flat-square&logo=jira&logoColor=0052cc)](https://myproject.atlassian.net/browse/XX-03)[![XX-04](https://img.shields.io/badge/-XX--04-dfe1e5.svg?style=flat-square&logo=jira&logoColor=0052cc)](https://myproject.atlassian.net/browse/XX-04) _(changelog)_ add GitHub and Jira badges to CHANGELOG
``` 
_which renders as_
> ### Features
>
> - [![b56ea](https://img.shields.io/badge/-b56ea-%23121011.svg?style=flat-square&logo=github&logoColor=white)](https://github.com/brianburwell11/cz-github-jira-conventional-footer/commit/b56ea677af8a969614b159316d9d28a3267f7169)[![XX-03](https://img.shields.io/badge/-XX--03-dfe1e5.svg?style=flat-square&logo=jira&logoColor=0052cc)](https://myproject.atlassian.net/browse/XX-03)[![XX-04](https://img.shields.io/badge/-XX--04-dfe1e5.svg?style=flat-square&logo=jira&logoColor=0052cc)](https://myproject.atlassian.net/browse/XX-04) _(changelog)_ add GitHub and Jira badges to CHANGELOG

## License

Distributed under the MIT License. See `LICENSE` for more information.

## Acknowledgements
This plugin would not have been possible without the fantastic work from:
* [commitizen tools](https://github.com/commitizen-tools/commitizen)
* [conventional_JIRA](https://github.com/Crystalix007/conventional_jira)
* [cz_github_jira_conventional](https://github.com/apheris/cz-github-jira-conventional)


<!-- links -->
[commitizen]: https://github.com/commitizen-tools/commitizen
[conventional-commit]: https://www.conventionalcommits.org/en/v1.0.0-beta.4/
[semver]: https://semver.org/
[pre-commit]: https://pre-commit.com/
[cz-defaults-bump-pattern]: https://github.com/commitizen-tools/commitizen/blob/0497d62ef9460e5961d90d7ba35fbb4e90b8e345/commitizen/defaults.py#L67
[cz-defaults-bump-map]: https://github.com/commitizen-tools/commitizen/blob/0497d62ef9460e5961d90d7ba35fbb4e90b8e345/commitizen/defaults.py#L68

CHANGELOG.md
## v1.0.2 (2022-06-13)

### Bug Fixes

- [![0af1d](https://img.shields.io/badge/-0af1d-%23121011.svg?style=flat-square&logo=github&logoColor=white)](https://github.com/brianburwell11/cz-github-jira-conventional-footer/commit/0af1dddfcc28b4e781d1ffbae48e51b89f0954cb) _(build)_ remove third-party import of yaml in setup script

## v1.0.1 (2022-06-13)

### Bug Fixes

- [![df29c](https://img.shields.io/badge/-df29c-%23121011.svg?style=flat-square&logo=github&logoColor=white)](https://github.com/brianburwell11/cz-github-jira-conventional-footer/commit/df29cb406a4acdcd4978bf1492725dca4dbeff29) _(build)_ fix the link to the github repo in `setup.py`

## v1.0.0 (2022-06-13)

### Bug Fixes

- [![53997](https://img.shields.io/badge/-53997-%23121011.svg?style=flat-square&logo=github&logoColor=white)](https://github.com/brianburwell11/cz-github-jira-conventional-footer/commit/53997cee306c596191f52ce44976f338f0013e1c)[![XX-99](https://img.shields.io/badge/-XX--99-dfe1e5.svg?style=flat-square&logo=jira&logoColor=0052cc)](https://myproject.atlassian.net/browse/XX-99) add `re.MULTILINE` flag to find Jira footers
- [![33855](https://img.shields.io/badge/-33855-%23121011.svg?style=flat-square&logo=github&logoColor=white)](https://github.com/brianburwell11/cz-github-jira-conventional-footer/commit/338556f2d9bb1b19c1f347eef5428b641eb98782) correct invalid regex pattern in commit_parser

## v0.3.0 (2022-06-09)

### Refactor

- [![6c157](https://img.shields.io/badge/-6c157-%23121011.svg?style=flat-square&logo=github&logoColor=white)](https://github.com/brianburwell11/cz-github-jira-conventional-footer/commit/6c1576bdb23ae984f0c943ee1c39a60da37e806a)[![XX-42](https://img.shields.io/badge/-XX--42-dfe1e5.svg?style=flat-square&logo=jira&logoColor=0052cc)](https://myproject.atlassian.net/browse/XX-42)[![XX-123](https://img.shields.io/badge/-XX--123-dfe1e5.svg?style=flat-square&logo=jira&logoColor=0052cc)](https://myproject.atlassian.net/browse/XX-123) _(config)_ use `dict.get()` to handle defaults for optional keys
- [![82b37](https://img.shields.io/badge/-82b37-%23121011.svg?style=flat-square&logo=github&logoColor=white)](https://github.com/brianburwell11/cz-github-jira-conventional-footer/commit/82b375b41aba9f8c52a2843ebfef5b6c49f4b979)[![XX-19](https://img.shields.io/badge/-XX--19-dfe1e5.svg?style=flat-square&logo=jira&logoColor=0052cc)](https://myproject.atlassian.net/browse/XX-19) _(config)_ raise custom exceptions for invalid config settings
- [![205e5](https://img.shields.io/badge/-205e5-%23121011.svg?style=flat-square&logo=github&logoColor=white)](https://github.com/brianburwell11/cz-github-jira-conventional-footer/commit/205e5ef59239fba8274188e2a4795021e75f2021)[![XX-06](https://img.shields.io/badge/-XX--06-dfe1e5.svg?style=flat-square&logo=jira&logoColor=0052cc)](https://myproject.atlassian.net/browse/XX-06) convert config checks to EAFP

### Features

- [![94fe6](https://img.shields.io/badge/-94fe6-%23121011.svg?style=flat-square&logo=github&logoColor=white)](https://github.com/brianburwell11/cz-github-jira-conventional-footer/commit/94fe63d6d84897f2d36dfa4c0054c676db6cbfb9)[![XX-99](https://img.shields.io/badge/-XX--99-dfe1e5.svg?style=flat-square&logo=jira&logoColor=0052cc)](https://myproject.atlassian.net/browse/XX-99) _(config)_ add optional config for standard commitizen settings

## v0.2.0 (2022-06-09)

### Features

- [![b56ea](https://img.shields.io/badge/-b56ea-%23121011.svg?style=flat-square&logo=github&logoColor=white)](https://github.com/brianburwell11/cz-github-jira-conventional-footer/commit/b56ea677af8a969614b159316d9d28a3267f7169)[![XX-03](https://img.shields.io/badge/-XX--03-dfe1e5.svg?style=flat-square&logo=jira&logoColor=0052cc)](https://myproject.atlassian.net/browse/XX-03)[![XX-04](https://img.shields.io/badge/-XX--04-dfe1e5.svg?style=flat-square&logo=jira&logoColor=0052cc)](https://myproject.atlassian.net/browse/XX-04) _(changelog)_ add GitHub and Jira badges to CHANGELOG
- [![22ebe](https://img.shields.io/badge/-22ebe-%23121011.svg?style=flat-square&logo=github&logoColor=white)](https://github.com/brianburwell11/cz-github-jira-conventional-footer/commit/22ebe396d6e6a336513ecd508631747900b91092)[![XX-02](https://img.shields.io/badge/-XX--02-dfe1e5.svg?style=flat-square&logo=jira&logoColor=0052cc)](https://myproject.atlassian.net/browse/XX-02) _(config)_ add a configurable `jira_token`
- [![e0c29](https://img.shields.io/badge/-e0c29-%23121011.svg?style=flat-square&logo=github&logoColor=white)](https://github.com/brianburwell11/cz-github-jira-conventional-footer/commit/e0c29e515fd50acf4e4339556940dd8c0d6b75ac) remove required validator on Jira issues
- [![eacc3](https://img.shields.io/badge/-eacc3-%23121011.svg?style=flat-square&logo=github&logoColor=white)](https://github.com/brianburwell11/cz-github-jira-conventional-footer/commit/eacc3543ee30be575455633471720a7a33502553) _(message)_ separate breaking changes from other footers
- [![c78eb](https://img.shields.io/badge/-c78eb-%23121011.svg?style=flat-square&logo=github&logoColor=white)](https://github.com/brianburwell11/cz-github-jira-conventional-footer/commit/c78eb3499eb997bf9124c01a1b9b7990e0a7c682) add question for jira issues
- [![aa657](https://img.shields.io/badge/-aa657-%23121011.svg?style=flat-square&logo=github&logoColor=white)](https://github.com/brianburwell11/cz-github-jira-conventional-footer/commit/aa6575373cc235c976539b6f36f966e3b683e955) remove jira requirement in message scope

### Bug Fixes

- [![48ee9](https://img.shields.io/badge/-48ee9-%23121011.svg?style=flat-square&logo=github&logoColor=white)](https://github.com/brianburwell11/cz-github-jira-conventional-footer/commit/48ee9e8a032dde2dcebfdd8f2ed6c94465e9cca7)[![XX-00](https://img.shields.io/badge/-XX--00-dfe1e5.svg?style=flat-square&logo=jira&logoColor=0052cc)](https://myproject.atlassian.net/browse/XX-00) _(message)_ add `!` after the type (or scope) of breaking change commits

cz_github_jira_conventional.py
import os
import re
from typing import Any, Dict, List

from commitizen import defaults, git, config
from commitizen.cz.base import BaseCommitizen
from commitizen.cz.utils import multiple_line_breaker, required_validator
from commitizen.cz.exceptions import CzException

__all__ = ["GithubJiraConventionalFooterCz"]


def parse_subject(text):
    if isinstance(text, str):
        text = text.strip(".").strip()

    return required_validator(text, msg="Subject is required.")


class GithubJiraConventionalFooterCz(BaseCommitizen):
    bump_pattern = defaults.bump_pattern
    bump_map = defaults.bump_map
    commit_parser = defaults.commit_parser
    changelog_pattern = defaults.bump_pattern

    # Read the config file and check if required settings are available
    conf = config.read_cfg()
    if "jira_prefix" in conf.settings:
        jira_prefix = conf.settings["jira_prefix"]
        issue_multiple_hint = "42, 123"
    else:
        jira_prefix = ""
        issue_multiple_hint = "XZ-42, XY-123"
    if "jira_base_url" not in conf.settings:
        print(
            "Please add the key jira_base_url to your .cz.yaml|json|toml config file."
        )
        quit()
    if "github_repo" not in conf.settings:
        print("Please add the key github_repo to your .cz.yaml|json|toml config file.")
        quit()
    jira_base_url = conf.settings["jira_base_url"]
    github_repo = conf.settings["github_repo"]
    # if "change_type_map" not in conf.settings:
    change_type_map = {
        "feat": "Feat",
        "fix": "Fix",
        "refactor": "Refactor",
        "perf": "Perf",
    }

    def questions(self) -> List[Dict[str, Any]]:
        questions: List[Dict[str, Any]] = [
            {
                "type": "list",
                "name": "prefix",
                "message": "Select the type of change you are committing",
                "choices": [
                    {
                        "value": "fix",
                        "name": "fix: A bug fix. Correlates with PATCH in SemVer",
                    },
                    {
                        "value": "feat",
                        "name": "feat: A new feature. Correlates with MINOR in SemVer",
                    },
                    {"value": "docs", "name": "docs: Documentation only changes"},
                    {
                        "value": "style",
                        "name": (
                            "style: Changes that do not affect the "
                            "meaning of the code (white-space, formatting,"
                            " missing semi-colons, etc)"
                        ),
                    },
                    {
                        "value": "refactor",
                        "name": (
                            "refactor: A code change that neither fixes "
                            "a bug nor adds a feature"
                        ),
                    },
                    {
                        "value": "perf",
                        "name": "perf: A code change that improves performance",
                    },
                    {
                        "value": "test",
                        "name": (
                            "test: Adding missing or correcting " "existing tests"
                        ),
                    },
                    {
                        "value": "build",
                        "name": (
                            "build: Changes that affect the build system or "
                            "external dependencies (example scopes: pip, docker, npm)"
                        ),
                    },
                    {
                        "value": "ci",
                        "name": (
                            "ci: Changes to our CI configuration files and "
                            "scripts (example scopes: GitLabCI)"
                        ),
                    },
                ],
            },
            {
                "type": "input",
                "name": "scope",
                "message": (
                    f'JIRA issue number (multiple "{self.issue_multiple_hint}"). {self.jira_prefix}'
                ),
                "filter": self.parse_scope,
            },
            {
                "type": "input",
                "name": "subject",
                "filter": parse_subject,
                "message": (
                    "Write a short and imperative summary of the code changes: (lower case and no period)\n"
                ),
            },
            {
                "type": "input",
                "name": "body",
                "message": (
                    "Provide additional contextual information about the code changes: (press [enter] to skip)\n"
                ),
                "filter": multiple_line_breaker,
            },
            {
                "type": "confirm",
                "message": "Is this a BREAKING CHANGE? Correlates with MAJOR in SemVer",
                "name": "is_breaking_change",
                "default": False,
            },
            {
                "type": "input",
                "name": "footer",
                "message": (
                    "Footer. Information about Breaking Changes and "
                    "reference issues that this commit closes: (press [enter] to skip)\n"
                ),
            },
        ]
        return questions

    def parse_scope(self, text):
        """
        Require and validate the scope to be Jira ids.
        Parse the scope and add Jira prefixes if they were specified in the config.
        """
        if self.jira_prefix:
            issueRE = re.compile(r"\d+")
        else:
            issueRE = re.compile(r"\w+-\d+")

        if not text:
            return ""

        issues = [i.strip() for i in text.strip().split(",")]
        for issue in issues:
            if not issueRE.fullmatch(issue):
                raise InvalidAnswerError(f"JIRA scope of '{issue}' is invalid")

        if len(issues) == 1:
            return self.jira_prefix + issues[0]

        return required_validator(
            ",".join([self.jira_prefix + i for i in issues]),
            msg="JIRA scope is required",
        )

    def message(self, answers: dict) -> str:
        prefix = answers["prefix"]
        scope = answers["scope"]
        subject = answers["subject"]
        body = answers["body"]
        footer = answers["footer"]
        is_breaking_change = answers["is_breaking_change"]

        if scope:
            scope = f"({scope})"
        if body:
            body = f"\n\n{body}"
        if is_breaking_change:
            footer = f"BREAKING CHANGE: {footer}"
        if footer:
            footer = f"\n\n{footer}"

        message = f"{prefix}{scope}: {subject}{body}{footer}"

        return message

    def example(self) -> str:
        return (
            "fix: correct minor typos in code\n"
            "\n"
            "see the issue for details on the typos fixed\n"
            "\n"
            "closes issue #12"
        )

    def schema(self) -> str:
        return (
            "<type>(<scope>): <subject>\n"
            "<BLANK LINE>\n"
            "<body>\n"
            "<BLANK LINE>\n"
            "(BREAKING CHANGE: )<footer>"
        )

    def schema_pattern(self) -> str:
        PATTERN = (
            r"(build|ci|docs|feat|fix|perf|refactor|style|test|chore|revert|bump)"
            r"(\(\S+\))?!?:(\s.*)"
        )
        return PATTERN

    def info(self) -> str:
        dir_path = os.path.dirname(os.path.realpath(__file__))
        filepath = os.path.join(dir_path, "conventional_commits_info.txt")
        with open(filepath, "r") as f:
            content = f.read()
        return content

    def process_commit(self, commit: str) -> str:
        pat = re.compile(self.schema_pattern())
        m = re.match(pat, commit)
        if m is None:
            return ""
        return m.group(3).strip()

    def changelog_message_builder_hook(
        self, parsed_message: dict, commit: git.GitCommit
    ) -> dict:
        """add github and jira links to the readme"""
        rev = commit.rev
        m = parsed_message["message"]
        if parsed_message["scope"]:
            parsed_message["scope"] = " ".join(
                [
                    f"[{issue_id}]({self.jira_base_url}/browse/{issue_id})"
                    for issue_id in parsed_message["scope"].split(",")
                ]
            )
        parsed_message[
            "message"
        ] = f"{m} [{rev[:5]}](https://github.com/{self.github_repo}/commit/{commit.rev})"
        return parsed_message


class InvalidAnswerError(CzException):
    ...


discover_this = GithubJiraConventionalFooterCz


cz_github_jira_conventional_footer.py
import os
import re
from typing import Any, Dict, List, Optional

from commitizen import defaults, git, config
from commitizen.exceptions import CommitizenException, ExitCode
from commitizen.cz.base import BaseCommitizen
from commitizen.cz.utils import multiple_line_breaker, required_validator
from commitizen.cz.exceptions import CzException
from yaml.scanner import Scanner, ScannerError

__all__ = ["GithubJiraConventionalFooterCz"]

CONVENTIONAL_COMMIT_REGEX = (
    r"(?s)\A(?P<change_type>feat|fix|refactor|perf)(?:\((?P<scope>\w+)\))?(?P<breaking>!)?: (?P<subject>.[^\n]*)"
    r"(?:\n^\n^(?P<body>.*))?"
)
"""regular expression pattern with named groups for conventional commits v1.0.0-beta.4"""


class MissingConfigKey(CommitizenException):
    """Exception raised when a required config key is not defined."""

    exit_code = ExitCode.INVALID_CONFIGURATION

    def __init__(self, config_key: str):
        super().__init__(config_key)
        self.message = f"Missing required config key: {config_key}"


class InvalidConfigValue(CommitizenException):
    """Exception raised when a config value doesn't match a criteria."""

    exit_code = ExitCode.INVALID_CONFIGURATION

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)


class InvalidAnswerError(CzException):
    ...


class GithubJiraConventionalFooterCz(BaseCommitizen):
    commit_parser = CONVENTIONAL_COMMIT_REGEX

    # read the config file and replace default if setting is defined
    try:
        conf = config.read_cfg()
    except ScannerError as e:
        if e.problem == "mapping values are not allowed here":
            error_message = (
                "Raw mapping values not allowed.\n\n"
                + 'Are you trying to define a footer token containing ":"? If so, be sure to enclose your defined value in quotation marks. For example,\n'
                + '\tjira_token: "resolves-issue: "'
            )
            raise InvalidConfigValue(error_message)
        raise

    jira_prefix = conf.settings.get("jira_prefix", "")
    issue_multiple_hint = "XZ-42, XY-123"
    if jira_prefix:
        issue_multiple_hint = "42, 123"
    jira_token = str(conf.settings.get("jira_token", "Jira: "))
    if not (jira_token.endswith(": ") or jira_token.endswith(" #")):
        jira_token = f"{jira_token}: "
    bump_pattern = conf.settings.get("bump_pattern", defaults.bump_pattern)
    bump_map = conf.settings.get("bump_map", defaults.bump_map)
    changelog_pattern = conf.settings.get("changelog_pattern", bump_pattern)
    change_type_map = conf.settings.get(
        "change_type_map",
        {
            "feat": "Features",
            "fix": "Bug Fixes",
            "refactor": "Refactor",
            "perf": "Performance Improvements",
        },
    )
    try:
        jira_base_url = conf.settings["jira_base_url"]
        github_repo = conf.settings["github_repo"]
    except KeyError as e:
        raise MissingConfigKey(e)

    # validate format for config settings
    if re.fullmatch(r"^(?i)BREAKING( |-)CHANGE( #|: )?", jira_token):
        raise InvalidConfigValue(
            "jira_token cannot match regex ^(?i)BREAKING( |-)CHANGE( #|: )"
        )
    if not (re.fullmatch(r"^(?i)[\w-]+(?: #|: )$", jira_token)):
        raise InvalidConfigValue("jira_token must match regex ^(?i)[\w-]+( #|: )?$")

    def questions(self) -> List[Dict[str, Any]]:
        questions: List[Dict[str, Any]] = [
            {
                "type": "list",
                "name": "prefix",
                "message": "Select the type of change you are committing",
                "choices": [
                    {
                        "value": "fix",
                        "name": "fix: A bug fix. Correlates with PATCH in SemVer",
                    },
                    {
                        "value": "feat",
                        "name": "feat: A new feature. Correlates with MINOR in SemVer",
                    },
                    {"value": "docs", "name": "docs: Documentation only changes"},
                    {
                        "value": "style",
                        "name": (
                            "style: Changes that do not affect the "
                            "meaning of the code (white-space, formatting,"
                            " missing semi-colons, etc)"
                        ),
                    },
                    {
                        "value": "refactor",
                        "name": (
                            "refactor: A code change that neither fixes "
                            "a bug nor adds a feature"
                        ),
                    },
                    {
                        "value": "perf",
                        "name": "perf: A code change that improves performance",
                    },
                    {
                        "value": "test",
                        "name": (
                            "test: Adding missing or correcting " "existing tests"
                        ),
                    },
                    {
                        "value": "build",
                        "name": (
                            "build: Changes that affect the build system or "
                            "external dependencies (example scopes: pip, docker, npm)"
                        ),
                    },
                    {
                        "value": "ci",
                        "name": (
                            "ci: Changes to our CI configuration files and "
                            "scripts (example scopes: GitLabCI)"
                        ),
                    },
                ],
            },
            {
                "type": "input",
                "name": "scope",
                "message": (
                    "What is the scope of this change? (class or file name): (press [enter] to skip)\n"
                ),
                "filter": parse_scope,
            },
            {
                "type": "input",
                "name": "subject",
                "filter": parse_subject,
                "message": (
                    "Write a short and imperative summary of the code changes: (lower case and no period)\n"
                ),
            },
            {
                "type": "input",
                "name": "body",
                "message": (
                    "Provide additional contextual information about the code changes: (press [enter] to skip)\n"
                ),
                "filter": multiple_line_breaker,
            },
            {
                "type": "confirm",
                "message": "Is this a BREAKING CHANGE? Correlates with MAJOR in SemVer",
                "name": "is_breaking_change",
                "default": False,
            },
            {
                "type": "input",
                "name": "breaking_change",
                "message": "Describe the breaking change:\n",
                "when": lambda x: x["is_breaking_change"],
            },
            {
                "type": "input",
                "name": "jira_issues",
                "message": (
                    f'JIRA issue number (multiple "{self.issue_multiple_hint}"). {self.jira_prefix}'
                ),
                "filter": self.parse_jira_issues,
            },
            {
                "type": "input",
                "name": "footer",
                "message": "Add any additional footers. (press [enter] to skip)\n",
            },
        ]
        return questions

    def parse_jira_issues(self, text):
        """
        Parse the issues given and add Jira prefixes if they were specified in the config.
        """

        if self.jira_prefix:
            issueRE = re.compile(r"\d+")
        else:
            issueRE = re.compile(r"\w+-\d+")

        if not text:
            return ""

        issues = [i.strip() for i in text.strip().split(",")]
        for issue in issues:
            if not issueRE.fullmatch(issue):
                raise InvalidAnswerError(f"Jira issue of '{issue}' is invalid")

        if len(issues) == 1:
            return self.jira_prefix + issues[0]

        return ",".join([self.jira_prefix + i for i in issues])

    def message(self, answers: dict) -> str:
        prefix = answers["prefix"]
        scope = answers["scope"]
        subject = answers["subject"]
        body = answers["body"]
        footer = answers["footer"]
        is_breaking_change = answers["is_breaking_change"]
        jira_issues = answers["jira_issues"]

        if scope:
            scope = f"({scope})"
        if body:
            body = f"\n\n{body}"
        if jira_issues:
            footer = f"{self.jira_token}{jira_issues}\n{footer}"
        if is_breaking_change:
            breaking_change = answers["breaking_change"]
            footer = f"BREAKING CHANGE: {breaking_change}\n{footer}"
        if footer:
            footer = f"\n\n{footer}"

        message = f"{prefix}{scope}{'!' if is_breaking_change else ''}: {subject}{body}{footer}"

        return message

    def example(self) -> str:
        return (
            "fix(lang)!: add polish language\n"
            "\n"
            "see the issue for details on the implementation\n"
            "\n"
            "BREAKING CHANGE: clients must add Polish as a language option\n"
            "Jira: XX-01"
        )

    def schema(self) -> str:
        return (
            "<type>(<scope>)(!): <subject>\n"
            "<BLANK LINE>\n"
            "<body>\n"
            "<BLANK LINE>\n"
            "(BREAKING CHANGE: <breaking change>)\n"
            "(<jira_token><jira issues>)\n"
            "<footer>"
        )

    def schema_pattern(self) -> str:
        PATTERN = (
            r"(build|ci|docs|feat|fix|perf|refactor|style|test|chore|revert|bump)"
            r"(\(\S+\))?!?:(\s.*)"
        )
        return PATTERN

    def info(self) -> str:
        dir_path = os.path.dirname(os.path.realpath(__file__))
        filepath = os.path.join(dir_path, "conventional_commits_info.txt")
        with open(filepath, "r") as f:
            content = f.read()
        return content

    def process_commit(self, commit: str) -> str:
        pat = re.compile(self.schema_pattern())
        m = re.match(pat, commit)
        if m is None:
            return ""
        return m.group(3).strip()

    def changelog_message_builder_hook(
        self, parsed_message: dict, commit: git.GitCommit
    ) -> dict:
        """add github and jira links to the readme as badges"""

        rev = commit.rev[:5]
        github_commit_badge_img = get_badge_image(
            "", rev, "%23121011", "github", "white"
        )
        github_commit_badge = f"[{github_commit_badge_img}](https://github.com/{self.github_repo}/commit/{commit.rev})"

        jira_issue_badges = []
        if parsed_message["body"]:
            jira_issues = []
            for jira_footer in re.findall(
                rf"^{self.jira_token}(?P<issues>.+)$",
                parsed_message["body"],
                re.IGNORECASE | re.MULTILINE,
            ):
                [
                    jira_issues.append(issue_id.strip())
                    for issue_id in jira_footer.split(",")
                ]

            for issue_id in jira_issues:
                issue_badge_img = get_badge_image(
                    "",
                    issue_id.replace("-", "--"),
                    "dfe1e5",
                    "jira",
                    "0052cc",
                    alt_text=issue_id,
                )
                issue_badge = (
                    f"[{issue_badge_img}]({self.jira_base_url}/browse/{issue_id})"
                )
                jira_issue_badges.append(issue_badge)

        scope = parsed_message["scope"]
        parsed_message["scope"] = None

        parsed_message["message"] = (
            f"{github_commit_badge}{''.join([badge for badge in jira_issue_badges])}"
            f"{f' _({scope})_' if scope is not None else ''} "
            f"{parsed_message['subject']}"
        )
        return parsed_message


def parse_scope(text):
    if not text:
        return ""

    scope = text.strip().split()
    if len(scope) == 1:
        return scope[0]

    return "-".join(scope)


def parse_subject(text):
    if isinstance(text, str):
        text = text.strip(".").strip()

    return required_validator(text, msg="Subject is required.")


def get_badge_image(
    label: str,
    value: str,
    background_color: str,
    logo: Optional[str] = None,
    logo_color: Optional[str] = None,
    style: Optional[str] = None,
    alt_text: Optional[str] = None,
) -> str:
    """get a badge from https://shields.io/ as markdown"""

    if not style:
        style = "flat-square"
    if not alt_text:
        alt_text = value
    badge_img = (
        f"https://img.shields.io/badge/-{value}-{background_color}.svg?style={style}"
    )
    if logo:
        badge_img = f"{badge_img}&logo={logo}"
        if logo_color:
            badge_img = f"{badge_img}&logoColor={logo_color}"
    badge_img = f"{badge_img}"
    return f"![{alt_text}]({badge_img})"


discover_this = GithubJiraConventionalFooterCz
